subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ExecutionStateChanges.java

```diff
@@ -16,32 +16,26 @@
 
 package org.gradle.internal.execution.history.changes;
 
+import com.google.common.collect.ImmutableSortedMap;
 import org.gradle.internal.change.Change;
 import org.gradle.internal.change.ChangeVisitor;
-import org.gradle.internal.execution.history.AfterPreviousExecutionState;
-
-import java.util.Optional;
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
 
 /**
  * Represents the complete changes in execution state
  */
 public interface ExecutionStateChanges {
 
-    int MAX_OUT_OF_DATE_MESSAGES = 3;
-
-    /**
-     * Returns changes to input files only, or {@link Optional#empty()} if a full rebuild is required.
-     */
-    Optional<Iterable<Change>> getInputFilesChanges();
-
     /**
      * Visits any change to inputs or outputs.
      */
     void visitAllChanges(ChangeVisitor visitor);
 
-    /**
-     * The base execution the changes are calculated against.
-     */
-    // TODO Use AfterPreviousExecutionState from context instead
-    AfterPreviousExecutionState getPreviousExecution();
+    <T> T visitInputFileChanges(IncrementalInputsVisitor<T> visitor);
+
+    interface IncrementalInputsVisitor<T> {
```

Comment:
I'd call this `InputChangesVisitor` -- there's no guarantee that it's going to be incremental, but it's surely going to be changes.

Reply:
That also got me thinking that the new interface should probably be called `InputFileChanges` or just `InputChanges` and not `IncrementalInputs`, given that you also get the type when the changes are not incremental.

Reply:
That makes a lot of sense to me. I like `InputChanges`, too.